tests/functional/base.py

```diff
@@ -0,0 +1,662 @@
+        console={
+            "class": "logging.StreamHandler",
+            "formatter": "bare",
+            "level": "DEBUG",
+            "stream": "ext://sys.stderr",
+        }
+    ),
+    loggers=dict(
+        test={
+            "level": "DEBUG",
+            "propagate": False,
+            "handlers": ["console"],
+        }
+    )
+)
+
+logging.config.dictConfig(LOGGING_CONF)
+logger = logging.getLogger('test')
+
+
+class BaseProviderTestSuite(unittest.TestCase):
+    """
+    Base test suite for a provider: docker, kubernetes, etc.
+    """
+
+    def setUp(self):
+        self.get_initial_state()
+
+    def tearDown(self):
+        self.restore_initial_state()
+
+    def get_initial_state(self):
+        raise NotImplementedError
+
+    def restore_initial_state(self):
+        raise NotImplementedError
+
+    def deploy(self, app_spec, answers):
+        """
+        Deploy to provider
+        """
+        raise NotImplementedError
+
+    def undeploy(self, app_spec, answers):
+        """
+        Undeploy from provider
+        """
+        raise NotImplementedError
+
+    def get_tmp_answers_file(self, answers):
+        f = tempfile.NamedTemporaryFile(delete=False, suffix='.conf')
+        f.close()
+        anymarkup.serialize_file(answers, f.name, format='ini')
+        return f.name
+
+    @property
+    def nulecule_lib(self):
+        return os.environ.get('NULECULE_LIB') or \
+            os.path.join(os.path.dirname(__file__), '../../../nulecule-library')
+
+    @classmethod
+    def disable_selinux(cls):
+        cls._enable_selinux = False
+        getenforce = subprocess.check_output('getenforce', shell=True).strip()
+        if getenforce == 'Enforcing':
+            subprocess.check_output('setenforce 0', shell=True)
+            cls._enable_selinux = True
+
+    @classmethod
+    def enable_selinux(cls):
+        if cls._enable_selinux:
+            subprocess.check_call('setenforce 1', shell=True)
+
+
+class DockerProviderTestSuite(BaseProviderTestSuite):
+    """
+    Base test suite for Docker.
+    """
+
+    def tearDown(self):
+        _containers = self._get_containers(all=True)
+
+        for container in _containers:
+            if container not in self._containers:
+                cmd = ['docker', 'rm', '-f', container]
+                print cmd
+                subprocess.check_output(cmd)
+
+    def deploy(self, app_spec, answers):
+        """
+        Deploy app to Docker
+
+        Args:
+            app_spec (str): image name or path to application
+            answers (dict): Answers data
+
+        Returns:
+            Path of the deployed dir.
+        """
+        destination = tempfile.mkdtemp()
+        answers_path = self.get_tmp_answers_file(answers)
+        cmd = ['atomicapp', 'run', '--answers=%s' % answers_path,
+               '--provider=docker',
+               '--destination=%s' % destination, app_spec]
+        subprocess.check_output(cmd)
+        return destination
```

Comment:
probably need to add function to download the latest atomic cli (or 1.8) and install to the system